File: frontend/src/api/ticket.service.ts
// Capa de API de Tickets (emisión + descarga de imagen/PDF).
import http from '@/api/http'
import type {
  CategoriaTicket,
  EmisionMasivaDto,
  FormatoPliego,
  ResumenImpresionDto,
  TicketDetalleDto,
} from '@/types/ticket.type'

/** Todos los tickets emitidos (activos). */
export function listarTickets() {
  return http.get<TicketDetalleDto[]>('/tickets/listar').then((r) => r.data)
}

export function emitirTicketEstudiante(idEstudiante: number) {
  return http
    .post<TicketDetalleDto>('/tickets/emitir-estudiante', null, { params: { idEstudiante } })
    .then((r) => r.data)
}

export function emitirTicketAdministrativo(idAdministrativo: number) {
  return http
    .post<TicketDetalleDto>('/tickets/emitir-administrativo', null, { params: { idAdministrativo } })
    .then((r) => r.data)
}

export function emitirTicketDocente(idDocente: number) {
  return http
    .post<TicketDetalleDto>('/tickets/emitir-docente', null, { params: { idDocente } })
    .then((r) => r.data)
}

/** Marca o desmarca un ticket como ENTREGADO (excluyente con rechazado). */
export function marcarEntrega(idTicket: number, entregado: boolean) {
  return http
    .patch<TicketDetalleDto>('/tickets/entrega', null, { params: { idTicket, entregado } })
    .then((r) => r.data)
}

/** Marca o desmarca un ticket como RECHAZADO / NO ACEPTO (excluyente con entregado). */
export function marcarRechazado(idTicket: number, rechazado: boolean) {
  return http
    .patch<TicketDetalleDto>('/tickets/rechazo', null, { params: { idTicket, rechazado } })
    .then((r) => r.data)
}

/** Cambia estado a PENDIENTE / ENTREGADO / RECHAZADO */
export function actualizarEstadoEntrega(idTicket: number, estado: 'ENTREGADO' | 'RECHAZADO' | 'PENDIENTE') {
  return http
    .patch<TicketDetalleDto>('/tickets/estado-entrega', null, { params: { idTicket, estado } })
    .then((r) => r.data)
}

// Solo el QR del ticket (para categorías sin plantilla aún, ej. administrativo).
export function obtenerTicketQr(idTicket: number) {
  return http.get(`/tickets/${idTicket}/qr`, { responseType: 'blob' }).then((r) => r.data as Blob)
}

// Los endpoints de imagen/PDF exigen el token JWT en el header, por eso NO se
// pueden usar en un <img src> directo: se piden con axios como 'blob'.
export function obtenerTicketPng(idTicket: number) {
  return http.get(`/tickets/${idTicket}/png`, { responseType: 'blob' }).then((r) => r.data as Blob)
}

export function obtenerTicketPdf(idTicket: number) {
  return http.get(`/tickets/${idTicket}/pdf`, { responseType: 'blob' }).then((r) => r.data as Blob)
}

/**
 * Emite tickets en lote. Sin argumento emite para TODOS los estudiantes activos;
 * con una lista de ids, solo para esos. Los que ya tenian ticket se omiten.
 */
export function emitirTicketsMasivo(idsEstudiante?: number[]) {
  return http
    .post<EmisionMasivaDto>('/tickets/emitir-estudiantes-masivo', idsEstudiante ?? null)
    .then((r) => r.data)
}

// --- Impresion por tandas ---

// En resumen y pliego, `carrera` es opcional: sin ella (o vacía) cuentan todas.
// Axios no manda los params que valen undefined, por eso el `|| undefined`.
export function resumenImpresion(
  formato: FormatoPliego = 'MIXTO_8',
  categoria: CategoriaTicket = 'ESTUDIANTE',
  carrera?: string,
) {
  return http
    .get<ResumenImpresionDto>('/tickets/impresion/resumen', {
      params: { formato, categoria, carrera: carrera || undefined },
    })
    .then((r) => r.data)
}

/**
 * Genera el PDF del pliego de UNA categoría (y de una carrera, si se indica). Por
 * defecto toma solo los pendientes y los deja marcados como impresos, para que la
 * próxima tanda siga donde quedó esta.
 */
export function generarPliego(opciones: {
  formato: FormatoPliego
  categoria: CategoriaTicket
  carrera?: string
  cantidad?: number
  soloPendientes?: boolean
  marcar?: boolean
  /** Ids de ticket en el orden en que deben salir (el de la tabla). Va en el cuerpo. */
  orden?: number[]
}) {
  return http
    .post('/tickets/impresion/pliego', opciones.orden?.length ? opciones.orden : null, {
      params: {
        formato: opciones.formato,
        categoria: opciones.categoria,
        carrera: opciones.carrera || undefined,
        cantidad: opciones.cantidad,
        soloPendientes: opciones.soloPendientes ?? true,
        marcar: opciones.marcar ?? true,
      },
      responseType: 'blob',
    })
    .then((r) => r.data as Blob)
}

export function marcarImpreso(idTicket: number, impreso: boolean) {
  return http.patch('/tickets/impresion/marcar', null, { params: { idTicket, impreso } })
}

/** Deja como NO impresos todos los tickets de una categoría (reinicia esa tanda). */
export function reiniciarImpresion(categoria: CategoriaTicket = 'ESTUDIANTE') {
  return http
    .post<number>('/tickets/impresion/reiniciar', null, { params: { categoria } })
    .then((r) => r.data)
}

/**
 * Actualiza por código adm: col2=materia, col3=SI/NO/RECHAZADO.
 * SI -> entregado (+ promueve), NO -> solo docente, RECHAZADO/NO ACEPTO -> rechazado.
 * Backend: POST /tickets/entrega-por-codigo?codigoAdm=&materia=&entrega=
 */
export function actualizarEntregaPorCodigo(codigoAdm: string, materia?: string, entrega?: string) {
  return http
    .post<TicketDetalleDto>('/tickets/entrega-por-codigo', null, {
      params: { codigoAdm, materia: materia || undefined, entrega: entrega || undefined },
    })
    .then((r) => r.data)
}

/**
 * Masivo: CSV con 3 columnas (codigo_adm, materia, SI/NO/RECHAZADO). Cada fila marca entrega
 * solo si col3=SI / rechazado si RECHAZADO, y promueve a docente si col2 trae dato.
 * Backend: POST /tickets/entrega-por-codigo/csv  multipart campo "archivo"
 */
export function actualizarEntregaPorCodigoCsv(archivo: File) {
  const fd = new FormData()
  fd.append('archivo', archivo)
  return http
    .post<import('@/types/estudiante.type').ImportacionResultadoDto>(
      '/tickets/entrega-por-codigo/csv',
      fd,
    )
    .then((r) => r.data)
}

// --- Estudiantes: marcar entregado por RU (1 columna) ---

export function marcarEntregaPorRu(ru: string, entrega?: string) {
  return http
    .post<TicketDetalleDto>('/tickets/entrega-por-ru', null, {
      params: { ru, entrega: entrega || undefined },
    })
    .then((r) => r.data)
}

export function marcarEntregaPorRuCsv(archivo: File) {
  const fd = new FormData()
  fd.append('archivo', archivo)
  return http
    .post<import('@/types/estudiante.type').ImportacionResultadoDto>(
      '/tickets/entrega-por-ru/csv',
      fd,
    )
    .then((r) => r.data)
}

File: src/main/java/com/uap/control_tickets/services/interfaces/TicketService.java
package com.uap.control_tickets.services.interfaces;

import com.uap.control_tickets.dto.ticket.EmisionMasivaDto;
import com.uap.control_tickets.dto.ticket.ResumenImpresionDto;
import com.uap.control_tickets.enums.CategoriaTicket;
import com.uap.control_tickets.enums.FormatoPliego;
import com.uap.control_tickets.dto.ticket.TicketDetalleDto;

import java.util.List;

/** Contrato de emisión y generación de tickets. */
public interface TicketService {
    List<TicketDetalleDto> listar();

    TicketDetalleDto obtener(Long idTicket);

    /**
     * Emite el ticket de un estudiante a partir de sus datos ya cargados.
     * Genera codigoIdentificacion + qrToken únicos. Si ya tenía ticket, lo devuelve.
     */
    TicketDetalleDto emitirEstudiante(Long idEstudiante);

    /** Emite el ticket de un administrativo (código ADM-… + qrToken). Idempotente. */
    TicketDetalleDto emitirAdministrativo(Long idAdministrativo);

    /** Emite el ticket de un docente (código DOC-… + qrToken). Idempotente. */
    TicketDetalleDto emitirDocente(Long idDocente);

    /**
     * Emite el ticket de varios estudiantes de una sola vez.
     * Si la lista es null o vacía se toman TODOS los estudiantes activos.
     * Como emitirEstudiante() es idempotente, los que ya tenían ticket se cuentan
     * como "omitidos" y no se duplican.
     */
    EmisionMasivaDto emitirEstudiantesMasivo(List<Long> idsEstudiante);

    // --- Impresion por tandas (cada categoria se imprime por separado) ---

    /**
     * Estado de impresion (impresos/pendientes/hojas) ACOTADO a una categoria y,
     * si se indica, a una carrera (solo aplica a ESTUDIANTE; null o vacio = todas).
     */
    ResumenImpresionDto resumenImpresion(FormatoPliego formato, CategoriaTicket categoria, String carrera);

    /**
     * Arma el PDF del proximo pliego de UNA categoria, acomodando los tickets en hojas oficio.
     *
     * @param formato   disposicion y medidas del ticket en la hoja.
     * @param categoria que tickets imprimir (solo se mezclan tickets de la misma categoria).
     * @param carrera   solo los estudiantes de esta carrera; null o vacio = todas.
     * @param cantidad  cuantos tickets incluir; null o <=0 = todos los pendientes.
     * @param soloPendientes true = toma solo los que nunca se imprimieron.
     * @param marcar    true = los deja marcados como impresos (para no repetirlos).
     * @param orden     ids de ticket en el orden en que deben salir (el de la tabla del
     *                  frontend); null o vacio = orden de emision. Solo cambia la POSICION
     *                  de cada ticket, no cuales entran.
     */
    byte[] generarPliego(FormatoPliego formato, CategoriaTicket categoria, String carrera,
                         Integer cantidad, boolean soloPendientes, boolean marcar, List<Long> orden);

    /** Marca o desmarca un ticket como impreso (por si hubo que reimprimir uno). */
    void marcarImpreso(Long idTicket, boolean impreso);

    /** Marca o desmarca un ticket como ENTREGADO (control de entrega física). */
    TicketDetalleDto marcarEntrega(Long idTicket, boolean entregado);

    /** Vuelve a dejar como no impresos todos los tickets de una categoria (reinicia esa tanda). */
    int reiniciarImpresion(CategoriaTicket categoria);

    /**
     * Busca por código administrativo, marca su ticket como ENTREGADO y,
     * si la segunda columna trae texto, lo promueve a DOCENTE con ese texto como carrera/materia.
     * La tercera columna (SI/NO) controla si se marca como entregado.
     *
     * Regla: SI -> marca entregado (+ promueve si hay materia); NO + materia -> solo promueve a docente, NO marca entregado.
     *
     * @param codigoAdm código administrativo (columna 1 del CSV)
     * @param materia   texto de la columna 2; si es null/vacío solo marca entrega, si trae dato convierte a docente
     * @param entregaFlag texto de la columna 3 (SI/NO); null/vacío = SI por compatibilidad; NO = no marca entrega
     */
    TicketDetalleDto actualizarPorCodigoAdm(String codigoAdm, String materia, String entregaFlag);

    /** Compatibilidad: 2 columnas (siempre marca entregado). */
    default TicketDetalleDto actualizarPorCodigoAdm(String codigoAdm, String materia) {
        return actualizarPorCodigoAdm(codigoAdm, materia, "SI");
    }

    /**
     * Variante masiva: CSV con 3 columnas por fila -> codigo_adm, materia, entrega(SI/NO).
     * Cada fila hace lo mismo que {@link #actualizarPorCodigoAdm}: promueve si hay materia y marca entrega solo si col3=SI.
     */
    com.uap.control_tickets.dto.estudiante.ImportacionResultadoDto actualizarPorCodigoAdmCsv(
            org.springframework.web.multipart.MultipartFile archivo);

    // --- Estudiantes: marcar entregado por RU (1 columna CSV) ---
    /** Marca como ENTREGADO el/los tickets del estudiante con ese RU. */
    TicketDetalleDto marcarEntregaPorRu(String ru);

    /** Marca como ENTREGADO el/los tickets del estudiante con ese RU, con flag SI/NO opcional. */
    TicketDetalleDto marcarEntregaPorRu(String ru, String entregaFlag);

    /** CSV con 1 columna (RU) -> marca entregado a cada estudiante listado. */
    com.uap.control_tickets.dto.estudiante.ImportacionResultadoDto marcarEntregaPorRuCsv(
            org.springframework.web.multipart.MultipartFile archivo);

    /** Ticket renderizado (con datos + QR) como PNG. */
    byte[] renderPng(Long idTicket);

    /** Ticket renderizado como PDF. */
    byte[] renderPdf(Long idTicket);
}
